docs/async/async-in-depth.md

```diff
@@ -0,0 +1,121 @@
+
+You can expect a server to be able to handle an order of magnitude more requests using `async` and `await` than if it were manually spawning threads to handle background jobs.
+
+### What does this mean for client scenario?
+
+The biggest gain for using `async` and `await` for a client app is an increase in responsiveness.  Although you can make an app responsive by spawning threads manually, the act of spawning a thread is an expensive operation relative to just using `async` and `await`.  Especially for something like a mobile game, impacting the UI thread as little as possible where I/O is concerned is crucial.
+
+More importantly, because the vast majority of wall clock time spent on I/O-bound code is effectively *doing nothing while the I/O happens*, dedicating an entire thread to perform barely any useful work is a poor use of resources.
+
+## Deeper Dive into Task and Task<T> for a CPU-Bound Operation
+
+Here's a 10,000 foot view of a CPU-bound async call:
+
+```csharp
+public async Task<int> CalculateResult(InputData data)
+{
+	// This queues up the work on the threadpool.
+	// This call is still synchronous!
+	var expensiveResultTask = Task.Run(() => DoExpensiveCalculation(data));
+	
+	// Note that at this point, you can do some other work concurrently.
+	
+	// Execution of CalculateResult is yielded here!
+	var result = await expensiveResultTask;
+	
+	return result;
+}
+```
+
+`CalculateResult` executes on the main thread.  When it calls `Task.Run`, it queues the expensive CPU-bound operation, `DoExpensiveCalculation`, on the thread pool and receives a `Task<int>` handle.  `DoExpensiveCalculation` is eventually run on the next available thread in the background.  A key takeaway is that the main thread hasn't stopped executing code yet.  It's possible to do concurrent work (such as updating a UI) while `DoExpensiveCalculation` is busy on another thread.
+
+Once `await` is encountered, the execution of `CalculateResult` is yielded to its caller, allowing other work to be done with the current thread while `DoExpensiveCalculation` is churning out a result.  Once it has finished, the result is queued up to run on the main thread.  Eventually, the main thread will return to executing `CalculateResult`, at which point it will have the result of `DoExpensiveCalculation`.
+
+**TODO:** Diagram!
+
+### Why does async help here?
+
+The use of `async` and `await` in conjunction with spawning a new task with `Task.Run` is worth considering.  Ultimately, it comes down to if you require responsiveness or not.
+
+If you *do** reuqire a responsive app or service, awaiting heavy CPU-bound work will be valuable.
+
+If you **do not** require a responsive app or service, there is nothing to gain by using `async` and `await`.  In fact, you will be doing things inefficiently because of the overhead involved in transforming `async` methods into state machines and queuing work across threads.
```

Comment:
Which APIs that are related to CPU-bound work are async?  AFAIK the only ones are for I/O-bound work.

I bring this up because there is overhead in the state machine the method is transformed into, and in scheduling the work to be done on another thread.  If the developer doesn't gain anything from responsiveness, then using `async` and `await` will be a performance hit over just doing the work synchronously and on the same thread.  I feel like that should be called out so that nobody gets the impression that `async` is the best way to do CPU-bound work.